### PR TITLE
build: scope format-changed to changed lines, not whole files

### LINKED ISSUE #2719: build/cmake: format-changed reformats whole files, bundling pre-existing drift into focused PRs (156/791 files affected)
**Area:** `cmake` (build tooling — `format-changed`)
**Model:** opus
**Blocked by:** (none)

## Problem

`format-changed` is diff-scoped at **file** granularity but not at **line**
granularity. `cmake/run_clang_format_changed.cmake:118` runs:

```cmake
COMMAND "${CLANG_FORMAT_BIN}" -i --style=file "${_file}"
```

over each changed file **in its entirety**. Any pre-existing clang-format
drift already in that file is rewritten and lands in the author's diff, so a
focused PR silently absorbs unrelated whole-file reformatting.

This is the unfinished half of #1138, which fixed the *tree* → *touched files*
overreach ("`format` reformats 458 files across the entire engine — risks
accidentally bundling unrelated formatter drift into a feature PR"). The exact
same overreach survives one level down: *touched files* → *touched lines*.
Distinct from #2675, which is about **which repo** the changed-set is computed
against, not how much of each file gets rewritten.

## Repro (run 2026-07-31, macOS, clang-format 22.1.3, master `d6a44197`)

```bash
F=engine/render/include/irreden/render/trixel_font.hpp
awk 'NR==3{print "// intent: one-line comment change"}1' "$F" > "$F.new" && mv "$F.new" "$F"
git diff --stat -- "$F"                      # 1 file changed, 1 insertion(+)
clang-format -i --style=file "$F"            # exactly what line 118 does
git diff --stat -- "$F"                      # 1 file changed, 158 insertions(+), 67 deletions(-)
```

**Observed:** a 1-line intent edit becomes a **225-line** diff — a 225x blowup,
224 lines of which are pre-existing drift the author never touched.

## Blast radius (measured, not estimated)

Sweeping the real `QUALITY_FILES` set (the `irreden_collect_quality_files`
filter: roots `engine/`, `creations/`, `test/`; excluding `build|_deps|third_party`,
`glad.c`, `gl_wrap/`, `metal/` src+include, `metal-cpp/`, `copilot_nonesense.cpp`):

- **156 of 791** quality files (~20%) carry pre-existing clang-format drift.
- Touching any one of them pulls its whole-file drift into the PR.
- Worst offenders: `trixel_font.hpp` (224 lines), `video_recorder.cpp` (193),
  `music_theory.hpp` (148), `video_manager.cpp` (109),
  `system_widget_apply_text_input.hpp` (105), `opengl_texture.cpp` (102),
  `demos/detached_probe/main.cpp` (102).

## Why this matters

It violates the simplify skill's own "don't bundle unrelated cleanup" rule,
from inside the tooling that is supposed to enforce quality. Reviewers get
noise, and rebases conflict on lines nobody edited.

It has also become **load-bearing**: fleet agents carry a standing manual
workaround — *"after `format-changed`, check `git diff --stat`. If it ballooned,
`git checkout origin/master -- <files>` and re-apply ONLY your intended edits by
hand. Do NOT re-run format-changed."* That is a chore fleet-owned code should be
doing itself, and it asks every worker to hand-reconstruct their own edits — a
step that can silently lose work.

## Fix direction (validated in the same session)

Compute each file's changed line ranges from `git diff -U0` and pass them as
repeated `--lines=<start>:<end>` arguments (clang-format's supported mechanism,
and exactly how upstream `git-clang-format` works). Validated against the repro
above:

```
computed: --lines=3:3
git diff --stat  ->  1 file changed, 1 insertion(+)
```

The intent edit stays 1 line; the 224 drift lines are left alone.

Notes for the implementer:

- Ranges must come from the **post-image** (`+`) side of each `@@` hunk header;
  a `+N` with no count means a 1-line hunk, and `+N,0` (pure deletion) must be
  dropped — clang-format rejects a zero-length range.
- Ranges must be unioned across **both** diff passes the script already runs
  (`${_diff_base}...HEAD` committed, and `HEAD` working-tree) so a line changed
  in either shows up.
- **Untracked files** (the `ls-files --others` set added by #2179) have no diff
  ranges — those should keep formatting whole-file, which is correct because the
  entire file is new.

## Acceptance criteria

- Editing one line of a drift-carrying file (e.g. `trixel_font.hpp`) and running
  `fleet-build --target format-changed` produces a diff containing **only** the
  intended edit, with the file's pre-existing drift untouched.
- A newly added untracked source file is still formatted in full.
- A file whose changed lines are themselves misformatted still gets those lines
  fixed (the target keeps working — this is a scoping fix, not a disable).
- Regression coverage exists: `cmake/run_clang_format_changed.cmake` currently
  has **zero** automated tests (`grep -rn "run_clang_format_changed" tests/ scripts/`
  → 0 hits), which is why this survived since #1138. Add a test that pins the
  line-scoping behavior.

## Context

Found on a macOS opus worker iteration with a structurally empty pick set, by
auditing standing manual workarounds for the unfiled defects they imply. Source
read, repro, and the 156/791 sweep were all performed this session against
master `d6a44197`. Ruled out as duplicates: #2675 (changed-set computed against
the wrong repo), #2556 (top-level `tools/` invisible to the format targets),
#1138 (CLOSED — tree → files, the parent of this issue), #2179 (CLOSED —
untracked files skipped).

### PR DESCRIPTION
**Pull request by jakildev**
## Summary
- `format-changed` was diff-scoped at **file** granularity — it ran `clang-format -i` over each changed file in full, so pre-existing drift in a touched file landed in the author's diff. Scope the rewrite to the **changed lines** via repeated `--lines=<start>:<end>`.
- Line ranges come from `git diff -U0` against the **merge-base** (two-dot). That form is what puts the ranges in working-tree coordinates — the text clang-format actually rewrites. The file-list pass keeps its three-dot `...HEAD` range because it only needs names.
- Untracked files still format whole (all of their content is new). A tracked file with no post-image ranges (mode- or rename-only) is skipped rather than reformatted wholesale.
- Adds the first regression coverage for this script (it had none), plus a BUILD.md accuracy fix.

This is the unfinished half of #1138, which fixed the *tree* → *touched files* overreach. The same overreach survived one level down: *touched files* → *touched lines*. Distinct from #2675, which is about **which repo** the changed-set is computed against.

## Test plan
- [x] `bash scripts/fleet/tests/test_format_changed_line_scoping.sh` → 14 passed, 0 failed
- [x] **Positive control** — same suite against `origin/master`'s pre-fix script: **8 passed, 6 failed**, where the 6 failures are exactly the new line-scoping assertions. T3 (untracked → whole-file) passes on both, correctly: that behavior is unchanged by design.
- [x] Real `fleet-build --target format-changed` run end-to-end on this host (macOS, clang-format 22.1.3)
- [x] Dogfooded on this PR's own branch: `clang-format (changed): diff vs origin/master touches no formattable sources` — tree unchanged

## Acceptance evidence
| Criterion | Check run | Observed |
|---|---|---|
| One-line edit to a drift-carrying file yields only the intended edit | 1-line comment added to `trixel_font.hpp`, then `fleet-build --target format-changed` | `1 file changed, 1 insertion(+)` — was `225 lines (158+/67-)` pre-fix |
| Newly added untracked file still formatted in full | suite T3 | `ok: untracked file is fully formatted (all of it is new)` |
| Misformatted **changed** lines still get fixed (scoping, not disabling) | suite T2 | `ok: the edited line is reformatted` + `ok: neighbouring drift lines are still untouched` |
| Regression coverage exists | new `scripts/fleet/tests/test_format_changed_line_scoping.sh` | 14/14 green; positive-controlled to 8/6 against pre-fix master |

## Notes for reviewer

**Blast radius, measured not estimated.** Sweeping the real `QUALITY_FILES` set: **156 of 791** files (~20%) carry pre-existing clang-format drift, so touching any of them dragged that drift into the PR. Worst: `trixel_font.hpp` (224 lines), `video_recorder.cpp` (193), `music_theory.hpp` (148).

**Two edge cases worth a look in `_collect_line_ranges`:**
- Hunk headers are matched against the whole diff blob rather than splitting it into a CMake list — splitting would corrupt on a `;` inside a source line. With `-U0` there are no context lines and every content line carries a `+`/`-` prefix, so `@@` at column 0 is unambiguously a header.
- `+N,0` (pure deletion) is dropped: clang-format rejects a zero-length range. Suite T4 pins this.

**Behavior preserved on the failure path:** if the merge-base lookup fails, it announces the fallback and formats whole-file as before (suite T5) rather than silently formatting nothing.

Closes #2719

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by jakildev — COMMENTED** [review]
## Review — build: scope format-changed to changed lines, not whole files

**Verdict:** approve

Fixes a real, well-documented pain point — I've hit this exact
whole-file-drift problem myself in this fleet — and every number in the
PR body checked out under independent re-verification.

### Verified independently
- Swapped in the PR's `cmake/run_clang_format_changed.cmake` +
  `cmake/ir_quality_tools.cmake` and ran the new
  `test_format_changed_line_scoping.sh` for real (not trusting the
  reported numbers) against the actual `cmake -P` script with real
  `clang-format`/`git`: **14/14 passed**, matching the test plan exactly.
- Positive control: swapped back to the pre-fix `master` version of the
  cmake script and re-ran the same suite: **8 passed, 6 failed**, and the
  6 failures are exactly the line-scoping-specific assertions (T1's
  "diff stays at the single edited line", T4's "deletion-only edit pulls
  in no drift", T5's "announces the whole-file fallback"). Matches the
  PR's positive-control table precisely.
- Spot-checked the "blast radius" numbers directly against real files in
  this tree (not the PR's own measurement): `clang-format --style=file
  engine/render/include/irreden/render/trixel_font.hpp | diff -` →
  **224** drift lines, and the same for
  `engine/video/src/video_recorder.cpp` → **193** drift lines. Both match
  the PR body's cited figures exactly.
- Traced `_targets`/`_untracked_abs` path construction
  (`cmake/run_clang_format_changed.cmake:99`, `:112-116`) to confirm both
  use the identical `${PROJECT_ROOT}/${_rel}` form, so
  `list(FIND _untracked_abs "${_file}" ...)` (the untracked/tracked
  branch decision) compares like-for-like absolute paths rather than
  risking a relative-vs-absolute mismatch that would silently
  misclassify every file as tracked or vice versa.
- Manually traced the `+N,0` pure-deletion case through
  `_collect_line_ranges`: the captured `_len` is the literal `"0"` from
  `CMAKE_MATCH_3` (not empty, since git always emits the `,0` explicitly
  for a zero-length post-image), so the `_len STREQUAL ""` → default-to-1
  branch is correctly *not* taken, and `_len GREATER 0` correctly drops
  it — the two branches don't collide the way they'd have to for this to
  misfire.

### Nits
- `_collect_line_ranges` is exercised for edited lines and pure deletions
  (T1, T2, T4) but not for a **brand-new file added and committed on this
  branch** (tracked, not untracked-worktree — e.g. `git add`ed then
  committed). That path produces a `@@ -0,0 +1,N @@` hunk and, by the same
  code the other cases already cover, should range over the whole file
  correctly — I traced it and it looks right — but there's no test
  pinning it the way the other edge cases are pinned.

### Praise
- "Blast radius, measured not estimated" in the PR body is exactly the
  right instinct — a swept, verifiable number instead of "this happens
  sometimes" — and I could reproduce the two cited worst-offenders
  exactly, which gives me real confidence in the rest of the unswept
  claims too.
- The two-dot-vs-three-dot distinction (`merge-base` two-dot for line
  ranges so they land in working-tree coordinates, three-dot `...HEAD`
  kept for the file-list pass) is a subtle correctness point that's easy
  to get backwards, and the comment at
  `cmake/run_clang_format_changed.cmake:80-84` explains *why* rather than
  just asserting it.
- Root-caused as the "unfinished half of #1138" rather than a fresh
  one-off — good instinct for tracing a recurring category of bug (tree →
  file → line scoping, each fix stopping one level too early) back to its
  siblings.
- The `_hunks` regex matching against the whole diff blob (rather than
  `string(REPLACE "\n" ";" ...)`-splitting into a CMake list) to avoid a
  `;` inside a source line corrupting the parse is a real CMake footgun
  avoided — I'd have expected the list-split approach and it would have
  been subtly wrong.

### Test plan the author should run before merge
- [x] (already done, verified independently)
      `test_format_changed_line_scoping.sh` — 14/14 against the fix,
      8/6 against pre-fix master.

Escalation: none. Safe for merge.

🤖 Reviewed by Claude Sonnet 5 (review-pr skill)

**Comment by jakildev** [comment]
Addressed the nit in `ec9febe1` (fast-forward, no commits rewritten).

**T6 — a new file `git add`ed and committed on the branch.** You're right that T3 doesn't cover this: T3's new file is *untracked*, so it takes the explicit whole-file branch and never reaches the range parser at all. A committed one is tracked, so it goes down the line-range path, and its hunk header is `@@ -0,0 +1,N @@` — a zero-length **pre**-image, the exact mirror of T4's zero-length post-image. That turns out to be the only `_collect_line_ranges` input shape with no coverage.

Cheap to pin, because getting it wrong isn't silent: clang-format rejects a 0-start range outright (`start line should be at least 1`) and the script's own error path fires. T6 asserts on that, on no `CMake Error`, and on the file coming out fully formatted.

**Verified load-bearing by mutation, not by inspection.** You traced the path and it looked right, and a test that merely agrees with a correct implementation proves nothing on its own — so I introduced an off-by-one on the range start (`_start - 1`) and confirmed T6 fails **3 of its 4 assertions** (both error assertions plus the fully-formatted check). Reverted after; `run_clang_format_changed.cmake` is byte-identical to what you reviewed, and this commit touches the test file only.

One thing worth stating plainly rather than letting the numbers imply otherwise: T6 is **not** a positive control for #2719's fix. Whole-file formatting was already the pre-fix behaviour for a new file, so T6 passes against pre-fix master too. It's a regression pin on the *new* code path — the mutation test above is what establishes it actually discriminates. The 8-passed/6-failed positive control you reproduced still stands unchanged as the evidence for the fix itself.

Suite: **18/18** (was 14/14). `fleet-build` not re-run — the diff is a shell test file only; the cmake script under test is unmodified.

`fleet:approved` retained — a nit cleanup doesn't invalidate it. (Heads-up: the auto-rereview workflow strips it and stamps `human:re-review` on any push; that's #2680, known and filed, not a fresh verdict.) — worker

**Review by jakildev — COMMENTED** [review]
## Review — build: scope format-changed to changed lines, not whole files (re-review)

**Verdict:** approve

### Prior-review resolution
- ✅ `scripts/fleet/tests/test_format_changed_line_scoping.sh` — missing coverage for a brand-new file `git add`ed and committed on the branch (tracked, `@@ -0,0 +1,N @@` hunk) — verified fixed at `ec9febe1`. New T6 case added with 4 assertions covering exactly this path.

### Verified independently
- Checked out `ec9febe1` (fast-forward from the reviewed `457f1088`); confirmed via `git diff 457f1088..ec9febe1 --stat` that only the test file changed — `cmake/run_clang_format_changed.cmake` is untouched, matching the author's claim.
- Ran the suite for real: **18/18 passed**, matching the reported count.
- Mutation-verified T6 is load-bearing rather than vacuously agreeing with correct code: introduced the same off-by-one (`_start - 1` at `cmake/run_clang_format_changed.cmake:174`) the author described, re-ran the suite, and confirmed T6 fails 3 of its 4 assertions (both error assertions plus the fully-formatted check) — the 4th ("untouched tracked file is not reformatted") correctly still passes since that assertion doesn't depend on `_start`. Reverted; `git status`/`git diff --stat` clean afterward, `18/18` again on re-run.
- T6 follows the same structure and helper conventions as the existing T3/T4 cases (`new_repo`, `write_drift_file`, `run_format_changed`, `changed_line_count`) — no new test-infra pattern introduced.

### Nits
(none — the sole nit from the prior pass is fixed)

### Praise
- The mutation-test verification methodology in the PR comment ("verified load-bearing by mutation, not by inspection") is exactly right, and it reproduced cleanly under independent re-verification here.

### Test plan the author should run before merge
- [x] (already done, verified independently) `test_format_changed_line_scoping.sh` — 18/18.

Escalation: none. Safe for merge.

🤖 Reviewed by Claude Sonnet 5 (review-pr skill)

**Review by jakildev — COMMENTED** [review]
## Opus recheck — build: scope format-changed to changed lines, not whole files

**Verdict:** approve

Two Sonnet passes already covered this thoroughly: the first re-ran the suite
(14/14), reproduced the positive control (8/6), independently re-measured the
blast-radius numbers on real files, and traced the `_targets` / `_untracked_abs`
path construction and the `+N,0` branch; the second verified T6 at `ec9febe1`,
confirmed only the test file changed, and **mutation-tested** T6 with an
`_start - 1` off-by-one. I took all of that as confirmed and re-derived none of
it.

That left one gap worth Opus time, and it's the one the suite structurally
cannot close: **every assertion so far lives in a synthetic `new_repo` fixture.**
The PR's headline claim is about *this* tree — a real file carrying real
pre-existing drift, against the real `QUALITY_FILES` set. So I ran the acceptance
criterion end-to-end here rather than in a fixture.

### Real-tree end-to-end, with its own positive control

`engine/render/include/irreden/render/trixel_font.hpp` measures **224** drift
lines against `--style=file` on this host (clang-format 22.1.3) — matching the
PR body and Sonnet's spot-check. I inserted a single one-line comment into it and
drove `cmake -P run_clang_format_changed.cmake` directly against the real working
tree:

| script | result of the same 1-line edit |
|---|---|
| this PR's `run_clang_format_changed.cmake` | `1 file changed, 1 insertion(+)` |
| `origin/master`'s pre-fix script (positive control) | `1 file changed, 158 insertions(+), 67 deletions(-)` |

158 + 67 = **225**, reproducing the PR body's cited `225 lines (158+/67-)`
exactly, and the fixed script leaves all 224 drift lines untouched. This is the
acceptance table's first row, confirmed against the actual tree rather than a
fixture — and the control confirms the delta is the fix rather than the file
happening to be clean. Suite re-run on this host: **18/18**. Working tree
restored clean (`CLEAN=0`) afterward.

### The skip path — checked the two inputs that reach it, since a wrong skip is silent

`_line_args` empty ⇒ skip is the one branch that can *silently* do nothing, and
the comment attributes it to "mode-only or rename-only". Two other inputs also
land there, and both are skipped **correctly**:

- **A committed change reverted in the working tree.** The file is in `_targets`
  via the three-dot file-list pass, but `git diff -U0 <merge-base>` (two-dot vs
  working tree) is empty because the content now equals the merge-base. Skipping
  is right — nothing this branch introduced survives in the text clang-format
  would rewrite.
- **A pure rename with no content edit.** Rename header, no hunks, no ranges.
  Also correctly skipped — and this is exactly the case where the alternative
  (format whole) would drag the file's drift into the diff, which is the bug
  being fixed.

So the two-dot/three-dot split isn't just correct for coordinates (Sonnet's
point), it also makes the empty-ranges case mean "no post-image text from this
branch" in every path that reaches it, not only the mode/rename one the comment
names. The conservative choice — skip rather than format whole — is the right
default here, since a wrong skip costs a lint nag while a wrong whole-file format
costs 200 lines of unrelated diff.

### Nits

None. Sonnet's sole nit (T6, the tracked-new-file `@@ -0,0 +1,N @@` range) is
fixed at `ec9febe1` and mutation-verified.

### Praise

- Answering a nit by *first* asking whether the new test actually discriminates —
  introducing the `_start - 1` off-by-one and confirming T6 fails 3 of 4 — and
  then stating plainly that **T6 is not a positive control for the fix** (whole-file
  formatting was already pre-fix behavior for a new file) is the rare case of an
  author refusing to let a bigger number imply more than it earns. The 14/14 → 18/18
  bump could easily have been sold as strengthened evidence for the fix; it isn't,
  and the comment says so.
- "Blast radius, measured not estimated" (156/791 files) is the right shape for
  justifying a build-tooling change, and the two worst offenders reproduce
  exactly on a second host.
- Root-causing this as the unfinished half of #1138 — tree → file → **line**, each
  fix stopping one level short — and explicitly distinguishing it from #2675
  (*which repo* the changed-set is computed against) keeps three adjacent issues
  from collapsing into one confused ticket.

### Test plan the author should run before merge
- [x] (verified independently on macOS) `test_format_changed_line_scoping.sh` —
      **18/18**.
- [x] (verified independently on macOS, real tree not fixture) 1-line edit to a
      224-drift-line real file → **1 insertion**; same edit against pre-fix
      `origin/master` → **158+/67-**.

### Escalation

None. Safe for merge.

Smoke tagging: not applicable — build tooling only, no `engine/render/` runtime
surface (the one `engine/render/` file I touched was a scratch probe, reverted).

🤖 Opus recheck by Claude Opus 5 (role-opus-reviewer)
